components/proxy/src/test/kotlin/com/hotels/styx/routing/handlers/LoadBalancingGroupTest.kt

```diff
@@ -0,0 +1,218 @@
+                frequencies["appx-05-b"]?.shouldBeGreaterThan(20)
+                frequencies["appx-06-c"]?.shouldBeGreaterThan(20)
+            }
+
+            scenario("... and emits NoAvailableHostsException when load balancing group is empty") {
+                routeDb.remove("appx-04")
+                routeDb.remove("appx-05")
+                routeDb.remove("appx-06")
+
+                routeDb.watch().waitUntil { it.entrySet().size == 2 }
+
+                // The lbGroup is still watching...
+                routeDb.watchers().shouldBe(1)
+
+                val e = shouldThrow<NoAvailableHostsException> {
+                    lbGroup.call(get("/").build())
+                }
+                e.message shouldBe "No hosts available for application appX"
+            }
+
+            scenario("... and exposes load balancing metric") {
+                routeDb.insert("appx-A", RoutingObjectRecord.create("HostProxy", setOf("appX"), mockk(), StaticResponseHandler(200, "appx-A")))
+                routeDb.insert("appx-B", RoutingObjectRecord.create("HostProxy", setOf("appX"), mockk(), StaticResponseHandler(200, "appx-B")))
+
+                routeDb.watch().waitUntil { it.entrySet().size == 4 }
+
+                val invocations = (1..40)
+                        .map {
+                            lbGroup.handle(get("/").build()).toMono()
+                        }
+
+                routeDb.get("appx-A").get().routingObject.metric().ongoingConnections() shouldBe 20
+                routeDb.get("appx-B").get().routingObject.metric().ongoingConnections() shouldBe 20
+
+                invocations.forEach {
+                    val response = it.block()
+                    println("response: ${response.bodyAs(UTF_8)}")
+                }
+            }
+
+            scenario("... stop load balancing group") {
+                routeDb.entrySet()
+                        .map { it.key }
+                        .forEach { routeDb.remove(it) }
+
+                lbGroup.stop()
+                routeDb.watchers() shouldBe 0
+            }
+        }
+
+        feature("Lifecycle handling") {
+            scenario("Stops watching route database") {
+                val watcher = mockk<Disposable>(relaxed = true)
+                val lbGroup = LoadBalancingGroup(mockk(), watcher)
+
+                lbGroup.stop()
+
+                verify {
+                    watcher.dispose()
+                }
+            }
+        }
+    }
+}
+
+fun Publisher<ObjectStore<RoutingObjectRecord>>.waitUntil(duration: Duration = Duration.ofSeconds(1), predicate: (ObjectStore<RoutingObjectRecord>) -> Boolean) = this
+        .toFlux()
+        .filter(predicate)
+        .blockFirst(duration)
+
+
+fun HttpHandler.call(request: HttpRequest, maxContentBytes: Int = 100000) = this.handle(request.stream(), HttpInterceptorContext.create())
+        .flatMap { it.aggregate(maxContentBytes) }
+        .toMono()
+        .block()
```

Comment:
```suggestion
        .block()

```
I think checkstyle complains about this. Though maybe it doesn't run on kotlin files.

Reply:
Are you referring to an empty line, or lack of?

Reply:
Our style says to end files with a newline.